lib/Page.js

```diff
@@ -694,6 +694,41 @@ class Page extends EventEmitter {
   }
 
   /**
+   * @param {string} selector
+   * @param {!Array<string>} values
+   */
+  async select(selector, ...values) {
+    const handle = await this.$(selector);
+    console.assert(handle, 'No node found for selector: ' + selector);
+
+    await this.evaluate((element, values) => {
+      if (element.nodeName.toLowerCase() !== 'select')
+        throw new Error('Element is not a <select> element.');
+
+      const options = Array.from(element.options);
+
+      if (Array.isArray(values) !== true)
+        values = [values];
+
+      if (element.multiple !== true) {
+        element.value = values[0];
+
+        if (element.value !== values[0])
```

Comment:
I'm concerned about falling back to the value of `label`. Depending on the combination of `value` and `label` attributes on every option, the matching will happen against either `value`, `label` or node's text content. The outcome is hard for me to predict.

Can we start with matching against values only? When/if there's a strong need to match against labels, I'd prefer it to be stated explicitly with options object.

Reply:
@aslushnikov For me this is a important feature, in fact for my specific use case I would specialize it even more. Often times, the option values have little resemblance to the label(s) you're targeting.

Nightmare.js has a convenient [`action`](https://github.com/segmentio/nightmare#extending-nightmare) method that can be used to implement custom behaviors in a global browser scope without repeating code and, AFAIK, puppeteer only has `exposeFunction` on a page scope.

Your `options` suggestion would still be limiting, what are your thoughts on having a `page.select(selector, callback)` signature, where the `callback` is a resolver function that gets the option `value`, `label`, `textContent`, `index` as arguments and has to return `true` to have that option selected, false otherwise?

A few scenarios I'm facing that could benefit from this (maybe applies to others too), given this:

```html
<select id="country">
	<option value="250">France</option>
	<option value="276">Deutschland</option>
	<option value="620">Portugal</option>
	<option value="724">España</option>
	<option value="840">United States of America</option>
</select>
```

You could select based on value directly:

```javascript
page.select('#country', '724');
```

Or implement advanced "resolvers", for instance ISO 3166-1 numeric to Alpha 2:

```javascript
page.select('#country', 'DE', (target, value, label, options) => {
  let resolved = null;

  switch (value) {
    case 250:
      resolved = 'FR';
    break;

    case 276:
      resolved = 'DE';
    break;

    case 620:
      resolved = 'PT';
    break;

    case 724:
      resolved = 'ES';
    break;

    case 840:
      resolved = 'US';
    break;
  }

  return target === resolved;
});
```

Or even go fuzzy if needed be:

```javascript
page.select('#country', 'Spain', (target, value, label, options) => {
  /**
   * Get all `options`, and find the one with the lowest
   * string distance from either the `value` or the `label`.
   *
   * In this case, "Spain" would be resolved to the value of "España".
   */
});
```

Reading other issues, this could also be helpful to resolve implementation inconsistencies between in libraries and frameworks (ReactJS Virtual DOM was mentioned for instance).

Reply:
> Nightmare.js has a convenient action method that can be used to implement custom behaviors in a global browser scope without repeating code and, AFAIK, puppeteer only has exposeFunction on a page scope.

If I understand Nightmare's action correctly, the analogue in puppeteer would be adding new methods to the page prototype:

```js
Page.prototype.size = function(done) {
  return this.evaluate(() => {
    const w = Math.max(document.documentElement.clientWidth, window.innerWidth || 0)
    const h = Math.max(document.documentElement.clientHeight, window.innerHeight || 0)
    return {
      height: h,
      width: w
    }
  });
})
```

It looks like it has nothing to do with `page.exposeFunction`.

> what are your thoughts on having a page.select(selector, callback) signature, where the callback is a resolver function that gets the option value, label, textContent, index as arguments and has to return true to have that option selected, false otherwise?

I think it's not a simple API.

If you can select by values, then you can implement custom resolver with `page.evaluate`. 
For example, selecting all options that contain letter `'e'` in their label:

```js
const values = await page.evaluate(() => {
  const options = Array.from(document.querySelectorAll('select > option'));
  const optionsWithE = options.filter(option => option.label.includes('e'));
  return optionsWithE.map(option => option.value);
  });
});
await page.select('select', ...values);
```

Reply:
@aslushnikov Extending the `Page` prototype would actually be super useful to me!

But I'm not sure how to go about accessing the `Page` class, it appears that it's not exported?

Reply:
@alixaxel ah right. You can "hack" it with a manual require:

```js
const Page = require('puppeteer/lib/Page');
```

This is not safe of course, but this magic require should not change much over the time.